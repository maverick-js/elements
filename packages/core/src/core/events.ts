import { type Dispose, onDispose } from '@maverick-js/signals';
import { anySignal, MaverickEventTarget, noop } from '@maverick-js/std';

import type { MaverickComponent } from './component';
import type { MaverickViewController } from './controller';
import type { TargetedEventHandler } from './types';

/** @internal */
export let $$_current_event_target: EventTarget | null = null;

const serverEventTarget: MaverickEventTarget<any> = /* #__PURE__ */ Object.freeze({
  dispatch: noop,
  dispatchEvent: noop,
  addEventListener: noop,
  removeEventListener: noop,
});

export function createEventTarget<Events>(): MaverickEventTarget<Events> {
  if (__SERVER__) return serverEventTarget;
  return ($$_current_event_target = new MaverickEventTarget<Events>());
}

/** @internal */
export function $$_set_current_event_target(target: EventTarget | null) {
  $$_current_event_target = target;
}

export type InferEvents<Target> =
  Target extends MaverickComponent<any, any, infer Events>
    ? Events
    : Target extends MaverickViewController<any, any, infer Events>
      ? Events
      : Target extends MaverickEventTarget<infer Events>
        ? Events extends {}
          ? Events
          : HTMLElementEventMap
        : Target extends { $ts__events?: infer Events }
          ? Events extends {}
            ? Events
            : HTMLElementEventMap
          : HTMLElementEventMap;

/**
 * Adds an event listener for the given `type` and returns a function which can be invoked to
 * remove the event listener.
 *
 * - The listener is removed if the current scope is disposed.
 * - This function is safe to use on the server (noop).
 */
export function listenEvent<
  Target extends EventTarget,
  Events = InferEvents<Target>,
  Type extends keyof Events = keyof Events,
>(
  target: Target,
  type: Type & string,
  handler: TargetedEventHandler<Target, Events[Type] extends Event ? Events[Type] : Event>,
  options?: AddEventListenerOptions | boolean,
): Dispose {
  if (__SERVER__) return noop;
  target.addEventListener(type, handler as any, options);
  return onDispose(() => target.removeEventListener(type, handler as any, options));
}

export class EventsController<Target extends EventTarget, Events = InferEvents<Target>> {
  #target: Target;
  #controller: AbortController;

  get signal(): AbortSignal {
    return this.#controller.signal;
  }

  constructor(target: Target) {
    this.#target = target;
    this.#controller = new AbortController();
    onDispose(this.abort.bind(this));
  }

  add<Type extends keyof Events>(
    type: Type,
    handler: TargetedEventHandler<Target, Events[Type] extends Event ? Events[Type] : Event>,
    options?: AddEventListenerOptions,
  ) {
    if (this.signal.aborted) throw Error('aborted');

    this.#target.addEventListener(type as any, handler as any, {
      ...options,
      signal: options?.signal ? anySignal(this.signal, options.signal) : this.signal,
    });

    return this;
  }

  remove<Type extends keyof Events>(
    type: Type,
    handler: TargetedEventHandler<Target, Events[Type] extends Event ? Events[Type] : Event>,
  ) {
    this.#target.removeEventListener(type as any, handler as any);
    return this;
  }

  abort(reason?: string) {
    this.#controller.abort(reason);
  }
}
