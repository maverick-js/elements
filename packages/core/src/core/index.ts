export * from './custom-element/attrs';
export * from './custom-element/types';
export * from './custom-element/internal';

export * from './component';
export * from './context';
export * from './controller';
export * from './events';
export * from './jsx';
export * from './instance';
export * from './lifecycle';
export * from './signals';
export * from './slots';
export * from './store';
export * from './symbols';
export * from './types';

export type { JSX } from '../jsx/jsx';
