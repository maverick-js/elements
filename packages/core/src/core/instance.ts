import { isFunction, MaverickEvent } from '@maverick-js/std';

import type { Component, ComponentConstructor, InferComponentProps } from './component';
import { provideContext } from './context';
import {
  ATTACH_SYMBOL,
  type AttachCallback,
  componentLifecycleSymbols,
  CONNECT_SYMBOL,
  type ConnectCallback,
  DESTROY_SYMBOL,
  type DestroyCallback,
  SETUP_SYMBOL,
  type SetupCallback,
} from './lifecycle';
import { createScope, getScope, onDispose, root, type Scope, scoped, signal } from './signals';
import { State as StateFactory } from './state';
import { ON_DISPATCH_SYMBOL } from './symbols';
import type { SignalOrValueRecord, WriteSignalRecord } from './types';

/** @internal */
export let $$_current_instance: AnyInstance | null = null;

export interface InstanceInit<Props = {}> {
  scope?: Scope | null;
  props?: Readonly<Partial<SignalOrValueRecord<Props>>> | null;
}

const EMPTY_PROPS = {} as any;

export interface AnyInstance extends Instance<any, any> {}

export class Instance<Props = {}, State = {}> {
  /* @internal */
  [ON_DISPATCH_SYMBOL]?: ((event: Event) => void) | null = null;

  readonly $host = signal<HTMLElement | null>(null);

  host: HTMLElement | null = null;
  scope: Scope | null = null;
  attachScope: Scope | null = null;
  connectScope: Scope | null = null;
  component: Component | null = null;
  destroyed = false;

  props: WriteSignalRecord<Props> = EMPTY_PROPS;

  state!: Readonly<State>;
  $state!: any;

  [SETUP_SYMBOL]: SetupCallback[] = [];
  [ATTACH_SYMBOL]: AttachCallback[] = [];
  [CONNECT_SYMBOL]: ConnectCallback[] = [];
  [DESTROY_SYMBOL]: DestroyCallback[] = [];

  constructor(scope: Scope, props: Props, state?: StateFactory<State>, init?: InstanceInit<Props>) {
    this.scope = scope;

    if (init?.scope) init.scope.append(scope);

    if (state) {
      this.$state = state.create();

      this.state = new Proxy(this.$state, {
        get: (_, prop: string) => this.$state[prop](),
      }) as State;

      provideContext(state, this.$state);
    }

    if (props) {
      this.props = createInstanceProps(props, init?.props) as WriteSignalRecord<Props>;

      if (init?.props) {
        for (const prop of Object.keys(init.props)) {
          const value = init.props[prop];
          if (!isFunction(value)) this.props[prop]?.set(value);
        }
      }
    }

    onDispose(this.destroy.bind(this));
  }

  setup() {
    $$_current_instance = this;

    const callbacks = this[SETUP_SYMBOL];
    if (callbacks) {
      scoped(() => {
        for (const callback of callbacks) callback();
      }, this.scope);
    }

    $$_current_instance = null;
  }

  attach(host: HTMLElement) {
    if (this.host) return;

    $$_current_instance = this;

    this.host = host;
    this.$host.set(host);

    if (__DEV__) {
      (host as any).$$COMPONENT_NAME = this.component?.constructor.name;
    }

    scoped(() => {
      this.attachScope = createScope();
      const callbacks = this[ATTACH_SYMBOL];
      if (callbacks.length > 0) {
        scoped(() => {
          for (const callback of callbacks) callback(this.host!);
        }, this.attachScope);
      }
    }, this.scope);

    this.component?.dispatchEvent(new MaverickEvent<void>('attach'));

    $$_current_instance = null;
  }

  detach() {
    this.attachScope?.dispose();
    this.attachScope = null;
    this.connectScope = null;

    if (__DEV__ && this.host) {
      (this.host as any).$$COMPONENT_NAME = null;
    }

    this.component?.dispatch(new MaverickEvent<void>('detach'));

    this.host = null;
    this.$host.set(null);
  }

  connect() {
    if (!this.host || !this.attachScope) return;

    $$_current_instance = this;

    scoped(() => {
      this.connectScope = createScope();
      const callbacks = this[CONNECT_SYMBOL];
      if (callbacks.length > 0) {
        scoped(() => {
          for (const callback of callbacks) callback(this.host!);
        }, this.connectScope);
      }
    }, this.attachScope);

    this.component?.dispatch(new MaverickEvent<void>('connect'));

    $$_current_instance = null;
  }

  disconnect() {
    this.connectScope?.dispose();
    this.connectScope = null;
    this.component?.dispatch(new MaverickEvent<void>('disconnect'));
  }

  destroy() {
    if (this.destroyed) return;
    this.destroyed = true;

    const callbacks = this[DESTROY_SYMBOL];
    if (callbacks.length > 0) {
      scoped(() => {
        for (const callback of callbacks) callback(this.host);
      }, this.scope);
    }

    const host = this.host;

    this.detach();
    this.scope!.dispose();

    for (const hook of Object.values(componentLifecycleSymbols)) {
      this[hook].length = 0;
    }

    this.props = EMPTY_PROPS;
    this.scope = null;
    this.state = EMPTY_PROPS;
    this.$state = null;

    // @ts-expect-error
    if (host) delete host.$;

    this.component?.dispatch(new MaverickEvent<void>('destroy'));
    this.component = null;
  }
}

function createInstanceProps<Props>(
  props: Props,
  init?: Partial<SignalOrValueRecord<Props>> | null,
) {
  const $props = {} as WriteSignalRecord<Props>;

  for (const name of Object.keys(props as Record<string, any>)) {
    const def = props[name];
    $props[name] = init && isFunction(init[name]) ? init[name] : signal(def, def);
  }

  return $props;
}

export function createComponent<T extends Component>(
  Component: ComponentConstructor<T>,
  init?: InstanceInit<InferComponentProps<T>>,
) {
  return root(() => {
    $$_current_instance = new Instance(getScope()!, Component.props, Component.state, init);

    const component = new Component();

    $$_current_instance.component = component;
    $$_current_instance = null;

    return component;
  });
}
