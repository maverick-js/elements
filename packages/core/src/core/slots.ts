import type { IsNever } from 'type-fest';

import type { JSX } from '../jsx/jsx';

// ---------------------------------------------------------------------------------------------
// createSlot
// ---------------------------------------------------------------------------------------------

export type Slot<Props = never> =
  IsNever<Props> extends true ? { (): JSX.Element } : { (props: Props): JSX.Element };

export type SlotChildren<Component> =
  Component extends Slot<infer Props>
    ? Props extends {}
      ? (props: Props) => JSX.Element
      : JSX.Element
    : never;

export type SlotProps<Name, Component> = Name extends 'default'
  ? { children: SlotChildren<Component> }
  : { name: Name; children: SlotChildren<Component> };

export type SlotComponent<Slots> = {
  <Name extends keyof Slots>(props: SlotProps<Name, Slots[Name]>): JSX.Element;
};

export function createSlot<Slots = { default: Slot }>(): SlotComponent<Slots> {
  return {} as any; // component is compiled out, return value is not used.
}

// ---------------------------------------------------------------------------------------------
// getSlots
// ---------------------------------------------------------------------------------------------

export interface DefaultSlots {
  default?: Slot;
}

export interface SlotRecord extends Record<string, Slot> {}

export type InferSlots<T> = T extends SlotComponent<infer Slots> ? Slots : T;

/** @internal */
export let $$_current_slots: SlotRecord = {};

export function getSlots<Slots = DefaultSlots>(slots?: Slots) {
  return $$_current_slots as InferSlots<Slots>;
}

/** @internal */
export function $$_set_current_slots(slots: SlotRecord) {
  $$_current_slots = slots;
}
