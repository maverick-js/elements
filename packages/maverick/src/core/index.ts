export * from './component';
export * from './context';
export * from './controller';
export * from './decorators';
export * from './instance';
export * from './signals';
export * from './state';
export * from './types';
