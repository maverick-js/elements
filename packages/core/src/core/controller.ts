import { MaverickEvent, type MaverickEventInit } from '@maverick-js/std';

import type {
  ComponentMeta,
  ResolveMetaEvents,
  ResolveMetaProps,
  ResolveMetaState,
} from './component';
import { $$_current_instance, Instance } from './instance';
import { type Scope, untrack } from './signals';
import { ON_DISPATCH_SYMBOL } from './symbols';
import type { ReadSignalRecord, WriteSignalRecord } from './types';

export class ViewController<
  Meta extends ComponentMeta = ComponentMeta,
  Props = ResolveMetaProps<Meta>,
  State = ResolveMetaState<Meta>,
  Events = ResolveMetaEvents<Meta>,
> extends EventTarget {
  /** @internal */
  readonly $$: Instance<Props, State>;

  get host(): HTMLElement | null {
    return this.$$.host;
  }

  get $host(): HTMLElement | null {
    return this.$$.$host();
  }

  get scope(): Scope {
    return this.$$.scope!;
  }

  get attachScope(): Scope | null {
    return this.$$.attachScope;
  }

  get connectScope(): Scope | null {
    return this.$$.connectScope;
  }

  get $props(): ReadSignalRecord<Props> {
    return this.$$.props;
  }

  get $state(): WriteSignalRecord<State> {
    return this.$$.store;
  }

  get state(): Readonly<State> {
    return this.$$.state;
  }

  constructor() {
    super();
    this.$$ = $$_current_instance!;
  }

  /**
   * Type-safe utility for creating component events.
   */
  createEvent<Type extends keyof Events = keyof Events>(
    type: Type & string,
    ...init: Events[Type] extends MaverickEvent
      ? Events[Type]['detail'] extends void | undefined | never
        ? [init?: Partial<MaverickEventInit<Events[Type]>['detail']>]
        : [init: MaverickEventInit<Events[Type]['detail']>]
      : never
  ): Events[Type] {
    return new MaverickEvent(type, init[0] as MaverickEventInit) as Events[Type];
  }

  /**
   * Creates a `MaverickEvent` and dispatches it. This method is typed to match all component events.
   */
  dispatch<Type extends Event | keyof Events>(
    type: Type,
    ...init: Type extends Event
      ? [init?: never]
      : Type extends keyof Events
        ? Events[Type] extends MaverickEvent
          ? Events[Type]['detail'] extends void | undefined | never
            ? [init?: Partial<MaverickEventInit<Events[Type]['detail']>>]
            : [init: MaverickEventInit<Events[Type]['detail']>]
          : [init?: never]
        : [init?: never]
  ): boolean {
    if (__SERVER__) return false;
    return this.dispatchEvent(
      type instanceof Event
        ? type
        : new MaverickEvent(type as string, init[0] as MaverickEventInit),
    );
  }

  override dispatchEvent(event: Event): boolean {
    if (__SERVER__) return false;
    return untrack(() => {
      this.$$[ON_DISPATCH_SYMBOL]?.(event);
      return super.dispatchEvent(event);
    });
  }
}
