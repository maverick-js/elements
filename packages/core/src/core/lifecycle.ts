import type { MaverickEvent } from '@maverick-js/std';

import { $$_current_instance } from './instance';

export const SETUP_SYMBOL = Symbol.for('maverick.setup');
export const ATTACH_SYMBOL = Symbol.for('maverick.attach');
export const CONNECT_SYMBOL = Symbol.for('maverick.connect');
export const DESTROY_SYMBOL = Symbol.for('maverick.destroy');
export const RENDER_SYMBOL = Symbol.for('maverick.render');

export const componentLifecycleSymbols = [
  SETUP_SYMBOL,
  ATTACH_SYMBOL,
  CONNECT_SYMBOL,
  DESTROY_SYMBOL,
];

export interface ComponentLifecycleEvents {
  /** Fired when the component attaches to a host element. */
  attach: MaverickEvent<void>;
  /** Fired when the component detaches from a host element. */
  detach: MaverickEvent<void>;
  /** Fired when the host element connects to the DOM. */
  connect: MaverickEvent<void>;
  /** Fired when the host element disconnects from the DOM. */
  disconnect: MaverickEvent<void>;
  /** Fired when the component instance is destroyed. */
  destroy: MaverickEvent<void>;
}

/**
 * The given callback is invoked when the component is ready to be set up.
 *
 * - This hook will run once.
 * - It's safe to use context inside this hook.
 * - The host element has not attached yet - wait for `onAttach`.
 */
export function onSetup(callback: SetupCallback) {
  $$_current_instance![SETUP_SYMBOL].push(callback);
}

export interface SetupCallback {
  (): void;
}

/**
 * The given callback is invoked when the component instance has attached to a host element.
 *
 * - This hook can run more than once as the component attaches/detaches from a host element.
 * - This hook may be called while the host element is not connected to the DOM yet.
 */
export function onAttach(callback: AttachCallback) {
  $$_current_instance![ATTACH_SYMBOL].push(callback);
}

export interface AttachCallback extends HostElementCallback {}

export interface HostElementCallback {
  (host: HTMLElement): any;
}

/**
 * The given callback is invoked when the host element has connected to the DOM.
 *
 * - This hook can run more than once as the host disconnects and re-connects to the DOM.
 */
export function onConnect(callback: ConnectCallback) {
  $$_current_instance![CONNECT_SYMBOL].push(callback);
}

export interface ConnectCallback extends HostElementCallback {}

/**
 * The given callback is invoked when the component is destroyed.
 *
 * - This hook will only run once when the component is finally destroyed.
 * - This hook may be called before being attached to a host element.
 * - This hook is called both client-side and server-side.
 */
export function onDestroy(callback: DestroyCallback) {
  $$_current_instance![DESTROY_SYMBOL].push(callback);
}

export interface DestroyCallback {
  (el: HTMLElement | null): void;
}
