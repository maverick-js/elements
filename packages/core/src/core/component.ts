import { isFunction } from '@maverick-js/std';
import type { WritableKeysOf } from 'type-fest';

import type { JSX } from '../jsx/jsx';
import { ViewController } from './controller';
import type { CustomElementOptions, MaverickCustomElement } from './custom-element/types';
import { Instance } from './instance';
import type { ComponentLifecycleEvents } from './lifecycle';
import { type Dispose, effect, type Maybe, scoped } from './signals';
import type { State } from './state';
import type { CUSTOM_ELEMENT_SYMBOL } from './symbols';
import type { SignalOrValueRecord } from './types';

const COMPONENT_CTOR_SYMBOL = /* #__PURE__ */ Symbol.for('maverick.component.ctor');

export class Component<Props = {}, State = {}, Events = {}, CSSVars = {}> extends ViewController<
  Props,
  State,
  Events & ComponentLifecycleEvents,
  CSSVars
> {
  static [COMPONENT_CTOR_SYMBOL] = true;

  /** @internal - DO NOT USE (for jsx types only) */
  jsxProps!: JSX.ComponentAttributes<Partial<SignalOrValueRecord<Props>>, Events, CSSVars>;

  render?(): JSX.Element;

  subscribe(callback: (state: Readonly<State>) => Maybe<Dispose>) {
    if (__DEV__ && !this.state) {
      const name = this.constructor.name;
      throw Error(
        `[maverick]: component \`${name}\` can not be subscribed to because it has no internal state`,
      );
    }

    return scoped(() => effect(() => callback(this.state)), this.$$.scope)!;
  }

  destroy(): void {
    this.$$.destroy();
  }
}

export function isComponentConstructor(value: unknown): value is ComponentConstructor {
  return isFunction(value) && COMPONENT_CTOR_SYMBOL in value;
}

export interface AnyComponent extends Component<any, any, any, any> {}

export interface ComponentConstructor<T extends Component = AnyComponent> {
  readonly element?: CustomElementOptions<InferComponentCSSProps<T>>;
  readonly props?: InferComponentProps<T>;
  readonly state?: State<InferComponentState<T>>;
  [CUSTOM_ELEMENT_SYMBOL]?(): MaverickCustomElement<T>;
  new (): T;
}

export type InferComponentProps<T> = T extends Component<infer Props> ? Props : {};

export type InferComponentState<T> = T extends Component<any, infer State> ? State : {};

export type InferComponentEvents<T> =
  T extends Component<any, any, infer Events> ? Events & ComponentLifecycleEvents : {};

export type InferComponentCSSProps<T> =
  T extends Component<any, any, any, infer CSSVars> ? CSSVars : {};

export type InferComponentMembers<T> =
  T extends Component<infer Props> ? Omit<Props, keyof T> & Omit<T, keyof Component> : {};

export type InferComponentCSSVars<
  Component extends AnyComponent,
  CSSProps = InferComponentCSSProps<Component>,
> = { [Var in WritableKeysOf<CSSProps> as `--${Var & string}`]: CSSProps[Var] };

export type InferComponentInstance<T> =
  T extends Component<infer Props, infer State> ? Instance<Props, State> : {};
