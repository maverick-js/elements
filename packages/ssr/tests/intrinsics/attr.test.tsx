import { signal } from '@maverick-js/core';

test('static', () => {
  expect(<div data-foo={1920} data-bar={1080} />).toMatchSnapshot();
});

test('dynamic', () => {
  const foo = 1920,
    bar = 1080;
  expect(<div data-foo={foo} data-bar={bar} />).toMatchSnapshot();
});

test('signal', () => {
  const $foo = signal(1920);
  expect(<div $data-foo={$foo} />).toMatchSnapshot();
});
