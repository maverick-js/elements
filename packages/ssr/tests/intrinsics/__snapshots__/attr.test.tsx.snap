// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dynamic 1`] = `"<!$><div data-foo="1920" data-bar="1080"></div>"`;

exports[`signal 1`] = `"<!$><div data-foo="1920"></div>"`;

exports[`static 1`] = `"<div data-foo="1920" data-bar="1080"></div>"`;
